SteamKit2/SteamKit2/Networking/Steam3/UdpConnection.cs

```diff
@@ -144,11 +146,18 @@ public override void Disconnect()
             if ( netThread == null )
                 return;
 
-            // Play nicely and let the server know that we're done. Other party is expected to Ack this,
-            // so it needs to be sent sequenced.
-            SendSequenced(new UdpPacket(EUdpPacketType.Disconnect));
+            // if we think we aren't already disconnected, apply disconnecting unless we read back disconnected
+            if ( state != (int)State.Disconnected && Interlocked.Exchange(ref state, (int)State.Disconnecting) == (int)State.Disconnected )

```

Comment:
state is advanced to State.Disconnected only when the thread is exiting. Every other case of advancing state is guarded with Interlocked.CompareExchange (to see that it was not advanced behind us). This Interlocked.Exchange is a case where we want to apply Disconnecting, unless the net thread just set Disconnected, in which case Disconnecting would never have advanced